### PR TITLE
Add FSDP2 + LoRA Fine-Tuning Pipeline/support for Orpheus TTS

### PR DESCRIPTION
**Pull request by MagellaX**
This PR introduces first-class support for fine-tuning Orpheus TTS with PyTorch FSDP2 and Low-Rank Adaptation (LoRA). The change set enables efficient multi-GPU training while preserving the model’s high-quality speech synthesis.

---

### Why

1. **Scalability:** Orpheus TTS runs close to the memory ceiling on a single GPU. FSDP2 shards parameters, gradients, and optimizer state, enabling larger batch sizes and faster wall-clock training.
2. **Parameter-Efficiency:** LoRA adds only a small number of trainable weights, cutting fine-tuning GPU memory by ~90 % compared with full fine-tuning and speeding convergence on modest data.
3. **Production Alignment:** Baseten, orpheus-cpp, and upcoming int8 inferences all benefit from smaller LoRA deltas and do not require full-weight storage.

---

### What

* **`finetune/fsdp2_lora_train.py`**  
  * End-to-end training script using FSDP2 + LoRA.  
  * Mixed-precision (bf16) and optional CPU offload.  
  * Activation checkpointing flag for further memory relief.  
  * Clean separation of dataset loading, model wrapping, and training loop; compatible with existing Orpheus token format (`input_ids`, `labels`, `attention_mask`).  

* **`finetune/fsdp2_lora_config.yaml`**  
  * Single source of truth for hyper-parameters, sharding strategy, LoRA ranks, checkpoint cadence, and logging.

* **`finetune/test_fsdp2_lora.py`**  
  * CI-friendly unit test that mocks a tiny Llama variant, runs one distributed step, and asserts loss propagation.

---

### How to Use

```bash
# update dataset_path in YAML, then:
torchrun --nproc_per_node=4 finetune/fsdp2_lora_train.py
```

The script saves merged LoRA checkpoints under `checkpoints/fsdp2-lora/` and supports automatic Weights & Biases logging.

---

### Backwards Compatibility

No existing inference or training paths are modified. All new files live under `finetune/` and are opt-in.

---

### Follow-Up Work that can be built on this PR...

1. **Float-8 and QLoRA:** Integrate 8-bit weight loading and FSDP2 float-8 all-gather for further memory and bandwidth savings.
2. **FSDP2 + Sequence Parallelism:** Add context-parallel sharding to push sequence length beyond 16 k tokens without increasing GPU count.
3. **Dataset Streamer:** Replace in-memory HF dataset with streaming WebDataset for scalable multilingual pretraining.
4. **Quantized Inference Export:** Provide a utility to merge and export LoRA-tuned weights as 4-bit GGML for orpheus-cpp.

Each of these can be delivered as an incremental, isolated PR building directly on this foundation.

### CODE REVIEW

**Comment by MagellaX** [comment]
Hey @amuvarma13, you can merge this

**Review by amuvarma13 — COMMENTED** [review]
Thanks for this I will definitely want to add something like this to the repo - wondering if there was a choice between using accelerator python api explicitly with an explicit training loop, vs huggingface trainer api.

**Comment by MagellaX** [comment]
> Thanks for this, I will definitely want to add something like this to the repo - wondering if there was a choice between using accelerator python api explicitly with an explicit training loop, vs huggingface trainer api.

Yeah, you could do that or use the TRL trainer; both will work. I chose a bare Accelerator loop because
1) FSDP-2 currently integrates more cleanly when you can control the wrapping points and sync calls;
2) the Orpheus data format (pre-tokenised input_ids / labels with masked text tokens) needs a custom loss step that’s easier to express in an explicit loop;
3) It keeps the door open for future tricks (float-8 all-gather, sequence-parallel sharding, streaming WebDataset) without fighting Trainer’s abstractions.
That said, the logic is modular: swap the loop for SFTTrainer/TRL and the rest of the plumbing (LoRA init, FSDP config) stays the same. 
Happy to refactor if ya'll guys prefer the higher-level API.

**Comment by MagellaX** [comment]
@EliasFiz any thoughts, bro?

**Comment by MagellaX** [comment]
> Thanks for this, I will definitely want to add something like this to the repo - wondering if there was a choice between using accelerator python api explicitly with an explicit training loop, vs huggingface trainer api.

BTW, you can merge this, it's all good!!

**Comment by MagellaX** [comment]
@amuvarma13 can you merge this bro!! 
I have to delete this branch afterwards

**Comment by amuvarma13** [comment]
> @amuvarma13 can you merge this bro!! I have to delete this branch afterwards

This is a great PR and we are keen to merge - we are currently testing and will merge/bug fix shortly.

**Comment by MagellaX** [comment]
Hey @amuvarma13 can I talk to you? Do you have an X account? Want to talk about hiring?

**Comment by amuvarma13** [comment]
> Hey @amuvarma13 can I talk to you? Do you have an X account? Want to talk about hiring?

Sure dm me https://x.com/akv_13

**Comment by MagellaX** [comment]
> > Hey @amuvarma13 can I talk to you? Do you have an X account? Want to talk about hiring?
> 
> Sure dm me https://x.com/akv_13

check!

**Comment by MagellaX** [comment]
Hey, are u going to merge this??
